### PR TITLE
Implement execute parameter of applyEdit command

### PR DESCRIPTION
**Pull request by 4e6**
[ci no changelog needed]

### Pull Request Description

PR implements new `execute` parameter designed in #3532. It allows IDE to control the execution after applying edits to files.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] ~If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide dist` and `./run ide watch`.~

### CODE REVIEW

File: engine/runtime-with-instruments/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala [hunk 1]
```diff
@@ -1031,6 +1032,111 @@ class RuntimeServerTest
...
+            TextEdit(
+              model.Range(model.Position(3, 8), model.Position(3, 9)),
+              "5"
+            )
+          ),
+          execute = false
+        )
+      )
+    )
+    context.receiveNone shouldEqual None
```
Comment (JaroslavTulach) [inline engine/runtime-with-instruments/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala:1117]: OK. There is a test for `execute = false`.

File: engine/runtime/src/main/scala/org/enso/interpreter/instrument/command/EditFileCmd.scala [hunk 2]
```diff
@@ -23,10 +24,13 @@ class EditFileCmd(request: Api.EditFileNotification) extends Command(None) {
     ctx.locking.acquireFileLock(request.path)
     ctx.locking.acquirePendingEditsLock()
     try {
-      ctx.state.pendingEdits.enqueue(request.path, request.edits)
+      val edits = request.edits.map(edit => PendingEdit(edit, request.execute))
+      ctx.state.pendingEdits.enqueue(request.path, edits)
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/scala/org/enso/interpreter/instrument/command/EditFileCmd.scala:28]: Rather then introducing `PendingEdit` one could have the original pending `TextEdit` and just a `Boolean` flag whether execution is needed. The processing of pending edits always is:
- apply all of them
- if one of them requested execution, perform execution at the end

E.g. we could got without `PendingEdit` class, I think.

File: engine/runtime/src/main/scala/org/enso/interpreter/instrument/execution/PendingEdits.scala [hunk 3]
```diff
@@ -11,12 +11,12 @@ trait PendingEdits {
     * @param file the edited file
     * @param edits the list of file edits
     */
-  def enqueue(file: File, edits: Seq[TextEdit]): Unit
+  def enqueue(file: File, edits: Seq[PendingEdit]): Unit
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/scala/org/enso/interpreter/instrument/execution/PendingEdits.scala:14]: This method would be `def enqueue(file: File, edits: Seq[TextEdit], execute: Boolean)`.

File: engine/runtime/src/main/scala/org/enso/interpreter/instrument/job/EnsureCompiledJob.scala [hunk 4]
```diff
@@ -237,13 +238,19 @@ final class EnsureCompiledJob(protected val files: Iterable[File])
...
     file: File
-  )(implicit ctx: RuntimeContext): Changeset[Rope] = {
+  )(implicit ctx: RuntimeContext): Option[Changeset[Rope]] = {
     ctx.locking.acquireFileLock(file)
     ctx.locking.acquireReadCompilationLock()
+    ctx.locking.acquirePendingEditsLock()
     try {
-      val edits = ctx.state.pendingEdits.dequeue(file)
-      ctx.executionService.modifyModuleSources(file, edits)
+      val pendingEdits = ctx.state.pendingEdits.dequeue(file)
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/scala/org/enso/interpreter/instrument/job/EnsureCompiledJob.scala:246]: `dequeue` would return `Pair[List[TextEdit], Boolean]`.

**Review by JaroslavTulach — APPROVED** [review]
I'd go without `PendingEdit` class, but that is just my personal preference. Otherwise OK.

File: engine/runtime/src/main/scala/org/enso/interpreter/instrument/command/EditFileCmd.scala (hunk 2 above)
Comment (4e6) [reply engine/runtime/src/main/scala/org/enso/interpreter/instrument/command/EditFileCmd.scala:28]: I don't see an issue with creating extra type. Imo, execute flag does not belong to the `TextEdit`, it's something external. Another reason I can think of, `TextEdit` and `PendingEdit` have different meaning and it's not a bad idea to create different types for them

**Pull request merged**